src/graphql/data/TokenPrice.ts

```diff
@@ -0,0 +1,173 @@
+import graphql from 'babel-plugin-relay/macro'
+import { useCallback, useEffect, useState } from 'react'
+import { fetchQuery } from 'react-relay'
+
+import type { Chain, TokenPriceAllQuery } from './__generated__/TokenPriceAllQuery.graphql'
+import type { HistoryDuration, TokenPriceSingleQuery } from './__generated__/TokenPriceSingleQuery.graphql'
+import TokenAPICache from './cache'
+import environment from './RelayEnvironment'
+import { TimePeriod } from './TopTokenQuery'
+import { PriceHistory } from './types'
+
+function toHistoryDuration(timePeriod: TimePeriod): HistoryDuration {
+  switch (timePeriod) {
+    case TimePeriod.HOUR:
+      return 'HOUR'
+    case TimePeriod.DAY:
+      return 'DAY'
+    case TimePeriod.WEEK:
+      return 'WEEK'
+    case TimePeriod.MONTH:
+      return 'MONTH'
+    case TimePeriod.YEAR:
+      return 'YEAR'
+    case TimePeriod.ALL:
+      return 'MAX'
+  }
+}
+
+const allQuery = graphql`
+  query TokenPriceAllQuery($contract: ContractInput!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory1H: priceHistory(duration: HOUR) {
+          timestamp
+          value
+        }
+        priceHistory1D: priceHistory(duration: DAY) {
+          timestamp
+          value
+        }
+        priceHistory1W: priceHistory(duration: WEEK) {
+          timestamp
+          value
+        }
+        priceHistory1M: priceHistory(duration: MONTH) {
+          timestamp
+          value
+        }
+        priceHistory1Y: priceHistory(duration: YEAR) {
+          timestamp
+          value
+        }
+        priceHistoryMAX: priceHistory(duration: MAX) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+const query = graphql`
+  query TokenPriceSingleQuery($contract: ContractInput!, $duration: HistoryDuration!) {
+    tokenProjects(contracts: [$contract]) {
+      name
+      markets(currencies: [USD]) {
+        priceHistory: priceHistory(duration: $duration) {
+          timestamp
+          value
+        }
+      }
+      tokens {
+        chain
+        address
+        symbol
+        decimals
+      }
+    }
+  }
+`
+
+export function useTokenPriceQuery(address: string, chain: Chain, timePeriod: TimePeriod) {
+  const [prices, setPrices] = useState<PriceHistory>(TokenAPICache.checkPriceHistory(address, timePeriod))
```

Comment:
so relay is designed to handle caching itself: https://medium.com/jusbrasil-tech/caching-fields-on-relay-modern-with-hooks-30d76502e404. i think unfortunately my understanding of the backend implementation is that ids are not stable, so we won't be able to use that. if you confirm that i think this solution works for now but just want you to be aware